message_processor.go

```diff
@@ -49,21 +52,28 @@ func (mg messageProcessor) consumeLog(lf *logPair) {
 	// create a protobuf reader for the log stream
 	dec := protoio.NewUint32DelimitedReader(lf.stream, binary.BigEndian, 1e6)
 	defer dec.Close()
+	defer lf.Close()
 	// a temp buffer for each log entry
 	var buf logdriver.LogEntry
+	curRetryNumber := 0
 	for {
-		// reads a message from the log stream and put it in a buffer until the EOF
-		// if there is any other error, recreate the stream reader
+		// reads a message from the log stream and put it in a buffer
 		if err := dec.ReadMsg(&buf); err != nil {
-			if err == io.EOF {
-				logrus.WithField("id", lf.info.ContainerID).WithError(err).Debug("shutting down log logger")
-				lf.stream.Close()
+			// exit the loop if reader reaches EOF or the fifo is closed by the writer or retry reaches the specified number
+			if err == io.EOF || err == os.ErrClosed || strings.Contains(err.Error(), "file already closed") || curRetryNumber >= mg.retryNumber {
+				logrus.WithField("id", lf.info.ContainerID).WithError(err).Info("shutting down log logger")
 				return
 			}
 
-			logrus.WithField("id", lf.info.ContainerID).WithError(err).Debug("Ignoring error")
-			dec = protoio.NewUint32DelimitedReader(lf.stream, binary.BigEndian, 1e6)
+			// if there is any other error, retry for robustness. If retryNumber is -1, retry forever
+			if curRetryNumber < mg.retryNumber || mg.retryNumber == -1 {
```

Comment:
Basically you will need return after the retry reach the max retries and you may like to first sleep and then init the `dec` object

```
if mg.retryNumber != -1 && curRetryNumber > mg.retryNumber {
     return
}
curRetryNumber += 1
logrus.WithField("id", lf.info.ContainerID).WithField("curRetryNumber", curRetryNumber).WithField("retryNumber", mg.retryNumber).WithError(err).Error("Encountered error and retrying")
time.Sleep(500 * time.Millisecond)
dec = protoio.NewUint32DelimitedReader(lf.stream, binary.BigEndian, 1e6)
```